ragna/source_storages/_lancedb.py

```diff
@@ -43,31 +43,54 @@ def __init__(self) -> None:
         import pyarrow as pa
 
         self._db = lancedb.connect(ragna.local_root() / "lancedb")
-        # Create schema at runtime!
-        self._schema = pa.schema(
-            [
-                pa.field("id", pa.string()),
-                pa.field("document_id", pa.string()),
-                pa.field("page_numbers", pa.string()),
-                pa.field("text", pa.string()),
-                pa.field(
-                    self._VECTOR_COLUMN_NAME,
-                    pa.list_(pa.float32(), self._embedding_dimensions),
-                ),
-                pa.field("num_tokens", pa.int32()),
-            ]
-        )
+
+        fields = [
+            pa.field("id", pa.string()),
+            pa.field("document_id", pa.string()),
+            pa.field("document_name", pa.string()),
+            pa.field("page_numbers", pa.string()),
+            pa.field("text", pa.string()),
+            pa.field(
+                self._VECTOR_COLUMN_NAME,
+                pa.list_(pa.float32(), self._embedding_dimensions),
+            ),
+            pa.field("num_tokens", pa.int32()),
+        ]
+
+        self._schema = pa.schema(fields)
 
     _VECTOR_COLUMN_NAME = "embedded_text"
 
+    def _get_table(self, corpus_name: Optional[str] = None):
+
+        table_name = corpus_name or self._embedding_id
+        return self._db.create_table(
+            name=table_name,
+            schema=self._schema,
+            exist_ok=True,
+        )
+
     def store(
         self,
         documents: list[Document],
         *,
+        corpus_name: Optional[str] = None,
         chunk_size: int = 500,
         chunk_overlap: int = 250,
     ) -> None:
-        table = self._db.create_table(name=self._embedding_id, schema=self._schema)
+
+        table = self._get_table(corpus_name)
+        all_fields = list(
+            set(table.schema.names)
+            | set([k for doc in documents for k in doc.metadata.keys()])
+        )
+        missing_fields = [
+            field for field in all_fields if field not in table.schema.names
+        ]
+
+        if missing_fields:
+            table.add_columns({field: "''" for field in missing_fields})
+            self._schema = table.schema
```

Comment:
Any reason they have to be lists? Keeping them as sets makes this easier and on top avoids the extra conversion.

```suggestion
        document_fields = {field for document in documents for field in document.metadata.keys()}
        schema_fields = set(table.schema.names)
        
        missing_fields = document_fields - schema_fields
        if missing_fields:
            table.add_columns({field: "''" for field in missing_fields})
            self._schema = table.schema

        extra_fields = schema_fields - document_fields
        extra_metadata = {field: "''" for field in extra_fields}
```